### DISCUSSION TITLE
Acoupi with Celery

### DISCUSSION THREAD

**Discussion by mbsantiago** [acoupi/acoupi#9] [Ideas]
# Using Celery in Acoupi

Hello @audevuilli, @ElliottRogers!

After researching Celery, I'm convinced it's a great fit for our project's needs. To make it easier to integrate Celery into Acoupi, I suggest we make some design changes.

## Acoupi Components

We've already implemented many Acoupi pieces that we can use as building blocks. However, to make the separation between components and the flows of the Acoupi processes clearer, I recommend grouping all components into a folder named components.

By doing so, we can easily identify the different pieces we need to assemble a process, and it will be easier to reuse components across different workflows. Additionally, this separation allows us to develop and maintain each component separately, making it easier to test and fix potential bugs.

## Acoupi Templates

To ease and standardize the creation of the Acoupi processes, I propose we create a module named templates. This module will take components as inputs and integrate them into the Acoupi processes required, such as recording, processing, and communication.

These templates will provide the building blocks that will be used in Celery tasks, allowing us to create workflows easily and quickly, while still providing flexibility and customization.

## Celery app

Once we have separated the app into components and templates, we can easily customize and integrate them with Celery. We can use the templates and components we need to build a Celery app that handles the execution and scheduling of tasks.

For example, we can use the templates.build_detection_process and templates.build_recording_process functions to create Celery tasks for recording audio and detecting bird calls.

By separating the app into components and templates, we can create workflows that are modular, flexible, and easy to test and maintain. Celery will provide the tools we need to manage tasks, schedule recurrent jobs, and scale up our app as needed.

Let me know what you think about these ideas and if you have any questions or concerns.

Cheers,
Santiago

**Comment by audevuilli** [comment]
Hi @mbsantiago 

Thanks for diving into Celery and highlighting the design changes. 

I created a new branch named `main-celery` to start making the design changes to implement celery. I would be happy not to have the changes on the current main branch, as I am using it to test the whole acoupi implementation on RPi (with mqtt, sqlite).

A few things came to mind when working on the system. Happy to hear your thoughts! 
1. Add a class in component recording_filter to save recordings according to **time_interval** (e.g. save recordings 1h after dusk and 1h before dawn, between 10pm and 2pm), to **time_sequence** (e.g. save recordings for 2 minutes every 30 minutes, every 1 hour), to **random** (e.g. save X recordings between start_time and end_time). 

    A. Could provide examples of how many days it will take to fill up the SD Card memory if x number of files are saved. Calculate this using the storage space available, duration of audio files at X.kHz, and number of files recorded per day.

2. Messengers.py is used to send data via MQTT. From my current understanding, we don't have a script at the moment that look into the SQLite database for "Failed message status" and send these message again. What should be the configuration for this script?

    A. As a function send_failed_message() within the MQTTMessenger class. The function will look into the sqlitedb for failed message status (e.g. constantly, every hour, every day, ...) and try again to send the failed message?
